src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -83,65 +88,61 @@ internal int FindMetricAggregators(ReadOnlySpan<KeyValuePair<string, object>> ta
-                    // The one created by this thread is wasted.
-                    this.keyValue2MetricAggs.TryGetValue(tagKey, out value2metrics);
+                    this.keyValue2MetricAggs.TryGetValue(seqKey, out value2metrics);
                 }
             }
 
             // GetOrAdd by TagValue at 2st Level of 2-level dictionary structure.
             // Get back Metrics[].
             if (!value2metrics.TryGetValue(tagValue, out aggregatorIndex))
             {
-                this.metricPointIndex++;
                 aggregatorIndex = this.metricPointIndex;
                 if (aggregatorIndex >= MaxMetricPoints)
                 {
                     // sorry! out of data points.
-                    // TODO: This doesn't account for "wasted space"
-                    // in the array. Need to reclaim them when we
-                    // really need. i.e inside this if block.
-                    // this can take explicit lock, as this
-                    // condition is not the common code path.
+                    // TODO: Once we support cleanup of
+                    // unused points (typically with delta)
+                    // we can re-claim them here.
                     return -1;
                 }
 
-                // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
-
-                if (seqKey == null)
-                {
-                    seqKey = new string[len];
-                    tagKey.CopyTo(seqKey, 0);
-                }
-
-                var seqVal = new object[len];
-                tagValue.CopyTo(seqVal, 0);
-                if (!value2metrics.TryAdd(seqVal, aggregatorIndex))
-                {
-                    // This thread lost the Add.
-                    // Some other thread added aggregatorIndex, take it.
-                    // aggregatorIndex incremented by this thread is
-                    // now a "wasted space" in the array.
-                    value2metrics.TryGetValue(tagValue, out aggregatorIndex);
-                    ref var metricPoint = ref this.metrics[aggregatorIndex];
-                    while (metricPoint.StartTime == default)
-                    {
-                        // spin wait until the thread who won adding to dictionary
-                        // also constructs a valid metric point.
-                    }
-                }
-                else
+                lock (value2metrics)
                 {
-                    // This thread won the Add.
-                    // Responsible for initializing MetricPoint.
-                    ref var metricPoint = ref this.metrics[aggregatorIndex];
-                    if (metricPoint.StartTime == default)
+                    // check again after acquiring lock.
+                    if (!value2metrics.TryGetValue(tagValue, out aggregatorIndex))
                     {
+                        this.metricPointIndex++;
```

Comment:
This needs to be an `Interlocked.Increment`. The lock taken here is by key & value, multiple threads could be executing this at the same time and ++ is not atomic. ++ does a read and a write (2 operations). Read is atomic, write is atomic, but they are not atomic together. 99% sure about this 😄

Reply:
Should be able to do it in one line too like: `aggregatorIndex = Interlocked.Increment(ref this.metricPointIndex)`

Reply:
yes you are right.. I think i initially used a different lock where it was okay to do plain ++. Now we need interlocked. Thanks!

Reply:
current test't wont catch it as they use same keys,values only. Need to write more complex tests here.

Reply:
tracking test additions here https://github.com/open-telemetry/opentelemetry-dotnet/issues/2372